### PR TITLE
socket: Respect IPv6 flowinfo and scope_id in InetAddr::from_std

### LINKED ISSUE #329: nix::sys::socket::InetAddr::from_std ignores sin6_flowinfo and sin6_scope_id
sin6_flowinfo and sin6_scope_id is essential for link-local addresses.
Related: carllerche/mio#376

### PR DESCRIPTION
**Pull request by kamalmarhubi**
Fixes #329

### CODE REVIEW

File: src/sys/socket/addr.rs [hunk 1]
```diff
@@ -36,12 +36,26 @@ pub enum InetAddr {
...
+                    .. unsafe { mem::zeroed() }
+                })
+            }
+            net::SocketAddr::V6(ref addr) => {
+                InetAddr::V6(libc::sockaddr_in6 {
+                    sin6_family: AddressFamily::Inet6 as sa_family_t,
+                    sin6_port: addr.port().to_be(),  // network byte order
+                    sin6_addr: Ipv6Addr::from_std(addr.ip()).0,
+                    sin6_flowinfo: addr.flowinfo().to_be(),  // network byte order
+                    sin6_scope_id: addr.scope_id().to_be(),  // network byte order
```
Comment (misuzu) [inline src/sys/socket/addr.rs]: sin6_flowinfo and sin6_scope_id must be in host byte order.
Related: rust-lang/rust#32424

Comment (kamalmarhubi) [inline src/sys/socket/addr.rs]: But the port should be in network byte order, right?

Comment (kamalmarhubi) [inline src/sys/socket/addr.rs]: From https://github.com/rust-lang/rust/pull/32429 looks like only the port. Should be fixed now!

**Comment by misuzu** [comment]
Looks great! Can't really test it as build is failing on nightly (nightly is required because of rust-lang/rust#32424).

**Comment by kamalmarhubi** [comment]
@misuzu it builds locally at 8422bfa on my machine. I think the CI is more strict with `raw_pointer_derive`. If you have time, could you check out that commit and see if it works?

**Comment by kamalmarhubi** [comment]
Actually, if you have a test case we can add, that would be even better!

**Comment by misuzu** [comment]
@kamalmarhubi std::net:SocketAddrV6 is broken on stable so test also will be broken. That stuff fixed in nightly 2016-03-25+
Seems like there is some bug in rustc.
```
$ rustc -V
rustc 1.9.0-nightly (b678600ac 2016-03-29)
$ cargo build
   Compiling nix v0.5.1-pre (file:///home/misuzu/Documents/Rust/nix-rust)
src/sys/signal.rs:157:9: 157:27 warning: lint raw_pointer_derive has been removed: using derive with raw pointers is ok, #[warn(renamed_and_removed_lints)] on by default
src/sys/signal.rs:157 #[allow(raw_pointer_derive)]
                              ^~~~~~~~~~~~~~~~~~
src/sys/signal.rs:157:9: 157:27 warning: lint raw_pointer_derive has been removed: using derive with raw pointers is ok, #[warn(renamed_and_removed_lints)] on by default
src/sys/signal.rs:157 #[allow(raw_pointer_derive)]
                              ^~~~~~~~~~~~~~~~~~
src/sys/signal.rs:158:23: 158:32 note: in this expansion of #[derive_PartialEq] (defined in src/sys/signal.rs)
src/sys/signal.rs:157:9: 157:27 warning: lint raw_pointer_derive has been removed: using derive with raw pointers is ok, #[warn(renamed_and_removed_lints)] on by default
src/sys/signal.rs:157 #[allow(raw_pointer_derive)]
                              ^~~~~~~~~~~~~~~~~~
src/sys/signal.rs:158:17: 158:21 note: in this expansion of #[derive_Copy] (defined in src/sys/signal.rs)
src/sys/signal.rs:157:9: 157:27 warning: lint raw_pointer_derive has been removed: using derive with raw pointers is ok, #[warn(renamed_and_removed_lints)] on by default
src/sys/signal.rs:157 #[allow(raw_pointer_derive)]
                              ^~~~~~~~~~~~~~~~~~
src/sys/signal.rs:158:10: 158:15 note: in this expansion of #[derive_Clone] (defined in src/sys/signal.rs)
src/sched.rs:208:20: 208:34 warning: `extern "C" fn(*mut Box<core::ops::FnMut() -> isize>) -> i32 {sched::clone::callback}` is now zero-sized and has to be cast to a pointer before transmuting to `*const extern "C" fn(*const Box<core::ops::FnMut() -> isize>) -> i32`, #[warn(transmute_from_fn_item_types)] on by default
src/sched.rs:208         ffi::clone(mem::transmute(callback), ptr as *mut c_void, flags.bits(), &mut cb)
                                    ^~~~~~~~~~~~~~
src/sched.rs:208:20: 208:34 warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
src/sched.rs:208:20: 208:34 note: for more information, see issue #19925 <https://github.com/rust-lang/rust/issues/19925>
rustc: /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/llvm/include/llvm/Support/Casting.h:237: typename llvm::cast_retty<X, Y*>::ret_type llvm::cast(Y*) [with X = llvm::Function; Y = llvm::Value; typename llvm::cast_retty<X, Y*>::ret_type = llvm::Function*]: Assertion `isa<X>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
error: Could not compile `nix`.

To learn more, run the command again with --verbose.
```

**Comment by misuzu** [comment]
I was wrong, this fix doesn't need rust-lang/rust#32424
You can add this test to test/sys/test_socket.rs
```rust
#[test]
pub fn test_inetv6_addr_to_sock_addr() {
    let port: u16 = 3000;
    let flowinfo: u32 = 1;
    let scope_id: u32 = 2;
    let actual = net::SocketAddr::V6(net::SocketAddrV6::new("fe80::1".parse().unwrap(), port, flowinfo, scope_id));
    let addr = InetAddr::from_std(&actual);

    match addr {
        InetAddr::V6(addr) => {
            assert_eq!(addr.sin6_port, port.to_be());
            assert_eq!(addr.sin6_flowinfo, flowinfo);
            assert_eq!(addr.sin6_scope_id, scope_id);
        }
        _ => panic!("nope"),
    }

    assert_eq!(actual, addr.to_std());
}
```

**Comment by kamalmarhubi** [comment]
Added the test, thanks!

**Comment by fiveop** [comment]
@homu r+

**Pull request merged**